sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4849,30 +4945,39 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
   // Lock automatically releases when this goes out of scope.
   std::scoped_lock QueueLock(Queue->Mutex);
 
-  _pi_ze_event_list_t TmpWaitList;
+  bool UseCopyEngine = false;
 
-  if (auto Res = TmpWaitList.createAndRetainPiZeEventList(NumEventsInWaitList,
-                                                          EventWaitList, Queue))
+  _pi_ze_event_list_t TmpWaitList;
+  if (auto Res = TmpWaitList.createAndRetainPiZeEventList(
+          NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine))
     return Res;
 
   // Get a new command list to be used on this call
   pi_command_list_ptr_t CommandList{};
   if (auto Res = Queue->Context->getAvailableCommandList(
-          Queue, CommandList, false /* UseCopyEngine */,
-          true /* AllowBatching */))
+          Queue, CommandList, UseCopyEngine, true /* AllowBatching */))
     return Res;
 
   ze_event_handle_t ZeEvent = nullptr;
-  pi_result Res = createEventAndAssociateQueue(
-      Queue, Event, PI_COMMAND_TYPE_NDRANGE_KERNEL, CommandList);
+  pi_event PiEvent;
+  pi_result Res = createEventAndAssociateQueue(Queue, &PiEvent,
+                                               PI_COMMAND_TYPE_NDRANGE_KERNEL,
+                                               CommandList, false, Event);
   if (Res != PI_SUCCESS)
     return Res;
-  ZeEvent = (*Event)->ZeEvent;
-  (*Event)->WaitList = TmpWaitList;
-
+  PiEvent->WaitList = TmpWaitList;
   // Save the kernel in the event, so that when the event is signalled
   // the code can do a piKernelRelease on this kernel.
-  (*Event)->CommandData = (void *)Kernel;
+  PiEvent->CommandData = (void *)Kernel;
+  if (Event) {
+    (*Event) = PiEvent;
+    ZeEvent = PiEvent->ZeEvent;
+  } else {
+    // Decrement the reference count of the event as this event will not be
+    // waited/released by SYCL RT, so it must be destroyed by EventRelease in
+    // resetCommandList.
+    PI_CALL(piEventRelease(PiEvent));
+  }
```

Comment:
I though we agreed this would be moved inside of `createEventAndAssociateQueue` since it be done each time after that call.